cdc/sinkv2/eventsink/txn/mysql/mysql.go

```diff
@@ -0,0 +1,728 @@
+		s.statistics.ObserveRows(event.Event.Rows...)
+	}
+
+	dmls := s.prepareDMLs()
+	log.Debug("prepare DMLs", zap.Any("rows", s.rows),
+		zap.Strings("sqls", dmls.sqls), zap.Any("values", dmls.values))
+
+	if err := s.execDMLWithMaxRetries(ctx, dmls); err != nil {
+		log.Error("execute DMLs failed", zap.String("err", err.Error()))
+		return errors.Trace(err)
+	}
+
+	s.events = s.events[:0]
+	s.rows = 0
+	return
+}
+
+func (s *mysqlSink) Close() (err error) {
+	if s.db != nil {
+		err = s.db.Close()
+		s.db = nil
+	}
+	if s.cancel != nil {
+		s.cancel()
+		s.cancel = nil
+	}
+	return
+}
+
+func (s *mysqlSink) MaxFlushInterval() time.Duration {
+	return maxFlushInterval
+}
+
+func querySQLMode(ctx context.Context, db *sql.DB) (sqlMode string, err error) {
+	row := db.QueryRowContext(ctx, "SELECT @@SESSION.sql_mode;")
+	err = row.Scan(&sqlMode)
+	if err != nil {
+		err = cerror.WrapError(cerror.ErrMySQLQueryError, err)
+	}
+	return
+}
+
+// check whether the target charset is supported
+func checkCharsetSupport(ctx context.Context, db *sql.DB, charsetName string) (bool, error) {
+	// validate charsetName
+	_, err := charset.GetCharsetInfo(charsetName)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	var characterSetName string
+	querySQL := "select character_set_name from information_schema.character_sets " +
+		"where character_set_name = '" + charsetName + "';"
+	err = db.QueryRowContext(ctx, querySQL).Scan(&characterSetName)
+	if err != nil && err != sql.ErrNoRows {
+		return false, cerror.WrapError(cerror.ErrMySQLQueryError, err)
+	}
+	if err != nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// prepareDMLs converts model.RowChangedEvent list to query string list and args list
+func (s *mysqlSink) prepareDMLs() *preparedDMLs {
+	startTsSet := make(map[uint64]struct{}, s.rows)
+	sqls := make([]string, 0, s.rows)
+	values := make([][]interface{}, 0, s.rows)
+	callbacks := make([]eventsink.CallbackFunc, 0, len(s.events))
+	replaces := make(map[string][][]interface{})
+
+	// flush cached batch replace or insert, to keep the sequence of DMLs
+	flushCacheDMLs := func() {
+		if s.params.batchReplaceEnabled && len(replaces) > 0 {
+			replaceSqls, replaceValues := reduceReplace(replaces, s.params.batchReplaceSize)
+			sqls = append(sqls, replaceSqls...)
+			values = append(values, replaceValues...)
+			replaces = make(map[string][][]interface{})
+		}
+	}
+
+	// translateToInsert control the update and insert behavior
+	translateToInsert := s.enableOldValue && !s.params.safeMode
+	for _, event := range s.events {
+		for _, row := range event.Event.Rows {
+			if !translateToInsert {
+				break
+			}
+			// It can be translated in to INSERT, if the row is committed after
+			// we starting replicating the table, which means it must not be
+			// replicated before, and there is no such row in downstream MySQL.
+			translateToInsert = row.CommitTs > row.ReplicatingTs
+		}
+	}
+
+	rowCount := 0
+	for _, event := range s.events {
```

Comment:
what's the relationship between the events and the rows? since the code execDMLWithMaxRetries(Line 498) will execute  the events in the same transaction , is it right ?